LFS-383 fixup

IntegralProgrammer opened the pull request:
Addresses the changes addressed in **LFS-383** but replaces the deletion and re-creation of answer sections (which breaks **LFS-273**) with a front-end fix that only deletes obsolete JCR nodes.

To test:
 - The test described on https://phenotips.atlassian.net/browse/LFS-383 should pass
 - The test described on https://phenotips.atlassian.net/browse/LFS-415 should pass
-  The test described on https://github.com/ccmbioinfo/lfs/pull/203#pullrequestreview-429826421 should pass

Additionally the following behavior should be observed:

 - Create a new *Demographics* form
 - Select **Yes** for *Undergoing surveillance*
 - **Save**
- In the JCR browser, the nodes corresponding to *Surveilled since* and *Surveilled until* should be flagged as *INCOMPLETE*
 - Enter a value for *Surveilled since*
 - **Save**
 - In the JCR browser, the status flags for *Surveilled since* should be empty
 - In the JCR browser, the status flags for *Surveilled until* should still be *INCOMPLETE*

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
WIP: Tested and seems to fix all problems associated with LFS-383, but needs code clean-up

IntegralProgrammer committed:
Removal of unreachable code

IntegralProgrammer committed:
Removal of debugging console.log statements

IntegralProgrammer committed:
Rename onConfigured to more meaningful answersTracker

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -120,11 +122,11 @@ let displaySection = (sectionDefinition, path, depth, existingAnswer, key) => {
  */
  export default function FormEntry(props) {
-  let { classes, entryDefinition, path, depth, existingAnswers, keyProp } = props;
+  let { classes, entryDefinition, path, depth, existingAnswers, keyProp, answersTracker, didGrow } = { ...{answersTracker: null, didGrow: null}, ...props };
```
```suggestion
  let { classes, entryDefinition, path, depth, existingAnswers, keyProp, answersTracker, didGrow } = props;
```
If we change the check in the comment below, it's possible to work around using null when you mean undefined

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -30,10 +30,18 @@ export const VALUE_POS = 1;
   let [ answerID ] = useState((existingAnswer && existingAnswer[0]) || uuidv4());
   let answerPath = path + "/" + answerID;
+  const [ isInitialized, setInitialized ] = useState(false);
+  if (isInitialized == false && answersTracker !== null) {
```
```suggestion
  if (isInitialized == false && answersTracker) {
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -30,10 +30,18 @@ export const VALUE_POS = 1;
 // for form submission
 function Answer (props) {
-  let { answers, answerNodeType, existingAnswer, path, questionName, questionDefinition, valueType, onChangeNote, noteComponent, noteProps } = props;
+  let { answers, answerNodeType, existingAnswer, path, questionName, questionDefinition, valueType, onChangeNote, noteComponent, noteProps, answersTracker, didGrow } = props;
```
Generally I feel our callbacks should be prefixed with "on" to conform with the usual HTML tags. So here, it's not "answersTracker", it should instead be named according to when it is called (onMount), and something similar for onGrow.

IntegralProgrammer replied:
I have renamed `didGrow` to `onAddedAnswerPath` to conform to the `onEvent` style of HTML. The other added property is `sectionAnswerState` which, I believe, should stay named as such as it is an array of length *2* with the `[0]` element being the dictionary state variable holding the mapping between question identifiers and lists of historic (and present) answer JCR paths, and the `[1]` element being the method to set the value of this dictionary.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -30,10 +30,18 @@ export const VALUE_POS = 1;
+  if (isInitialized == false && answersTracker !== null) {
+    if (!(answerPath in answersTracker[0])) {
+      answersTracker[1](answersTracker[0].concat(answerPath));
+      didGrow(true);
```
This creates a warning, since we're editing the state of a parent during rendering of its children, which is discouraged:
```
react_devtools_backend.js:2273 Warning: Cannot update a component (`Section`) while rendering a different component (`Answer`). To locate the bad setState() call inside `Answer`, follow the stack trace as described in https://fb.me/setstate-in-render
    in Answer (created by NumberQuestion)
    in div (created by ForwardRef(CardContent))
<snip a bunch of lines>
```

IntegralProgrammer replied:
@OrdiNeu I have created the commit `3ff5f42` to address this. I have done all the tests and everything seems to behave properly without generating any React warnings. Please let me know if the solution is good, if the warning is still generated, if new bugs are introduced, etc...

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -103,6 +105,28 @@ function Section(props) {
+  const totalAnswers = Object.entries(sectionDefinition).filter(([key, value]) => ENTRY_TYPES.includes(value['jcr:primaryType'])).length;
+  const answerStateVars = [];
+  for (let i = 0; i < totalAnswers; i++) {
+    let [ stateVar, stateVarSetter ] = useState([]);
```
While this works as-is while filling out forms, this can cause strange behaviour if totalAnswers ever changes between renders, since [React depends on the order in which useStates() are called to figure out what to return](https://reactjs.org/docs/hooks-faq.html#how-does-react-associate-hook-calls-with-components).

IntegralProgrammer replied:
This should be fixed by commit `988298c`

OrdiNeu reviewed (changes requested):
Works well, but some of the variables should be renamed.

IntegralProgrammer committed:
Apply suggestions from code review: change null to undefined

By checking if a function/variable is undefined instead of null, we can write cleaner code by avoiding the need to initialize the variable to null

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
Solution for warning generated when the state of the Section component was updated during the rendering of the Answer component

IntegralProgrammer committed:
Remove improper design React pattern which calls useState() in a loop with a single call to useState() holding a dictionary as a state variable mapping question identifiers to lists of historic (and present) answer JCR paths

Change name of didGrow callback to onAddedAnswerPath

sashaandjic reviewed (approved):
Tested all 4 suggested scenarios and all worked as expected.

OrdiNeu committed:
[wip] Adding code to prevent us needing a separate state variable just
to track answer changes.

OrdiNeu committed:
[wip] Fixed the initial value of the access count variable

sashaandjic reviewed (approved):
re-tested all 4 scenarios
all good
